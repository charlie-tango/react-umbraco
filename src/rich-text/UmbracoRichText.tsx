import { decode } from "html-entities";
import React from "react";
import {
  type RenderBlockContext,
  type RichTextElementModel,
  type RouteAttributes,
  isHtmlElement,
  isRootElement,
  isTextElement,
  isUmbracoBlock,
} from "./RichTextTypes";
import { parseStyle } from "./parse-style";

interface NodeMeta {
  /** The tag of the parent element */
  ancestor?: string;
  /** The tag of the previous sibling element */
  previous?: string;
  /** The tag of the next sibling element */
  next?: string;
}

/**
 * Props for rendering a single node in the rich text.
 * A node is any HTML element that is part of the rich text.
 */
export type RenderNodeContext = {
  children?: React.ReactNode;
  meta: NodeMeta;
} & (
  | {
      [Tag in keyof React.JSX.IntrinsicElements]: {
        tag: Tag;
        attributes: React.JSX.IntrinsicElements[Tag];
      };
    }[keyof Omit<React.JSX.IntrinsicElements, "a">]
  | {
      tag: "a";
      attributes: React.JSX.IntrinsicElements["a"];
      /** The route attributes for internal Umbraco links */
      route?: RouteAttributes;
    }
);

interface RichTextProps {
  data: RichTextElementModel | undefined;
  renderBlock?: (block: RenderBlockContext) => React.ReactNode;
  /**
   * Render an HTML node with custom logic.
   * @param node
   * @returns A React node, `null` to render nothing, or `undefined` to fallback to the default element
   */
  renderNode?: (node: RenderNodeContext) => React.ReactNode | undefined;
  /** Default attributes for HTML elements, used to add default classes to all `<p>` tags.
   * If the html element contains its own attributes, then they will override the default.
   *
   * ```tsx
   * <RichText
   *    htmlAttributes={{
   *      p: { className: 'text-base' },
   *      h1: { className: 'text-2xl' },
   *    }}
   *  />
   *  */
  htmlAttributes?: Partial<{
    [Tag in keyof React.JSX.IntrinsicElements]: React.JSX.IntrinsicElements[Tag];
  }>;
}

/**
 * Render the individual elements of the rich text
 */
function RichTextElement({
  element,
  blocks,
  renderBlock,
  renderNode,
  htmlAttributes = {},
  meta,
}: {
  element: RichTextElementModel;
  blocks: Array<RenderBlockContext> | undefined;
  meta:
    | {
        ancestor?: string;
        next?: string;
        previous?: string;
      }
    | undefined;
} & Pick<RichTextProps, "renderBlock" | "renderNode" | "htmlAttributes">) {
  if (!element || element.tag === "#comment" || element.tag === "#root")
    return null;

  if (isTextElement(element)) {
    // Decode HTML entities in text nodes
    return decode(element.text);
  }

  // If the tag is a block, skip the normal rendering and render the block
  if (isUmbracoBlock(element)) {
    const block = blocks?.find(
      (block) => block.content?.id === element.attributes?.["content-id"],
    );
    if (renderBlock && block) {
      return renderBlock(block);
    }
    if (typeof renderBlock !== "function") {
      throw new Error(
        "No renderBlock function provided for rich text block. Unable to render block.",
      );
    }

    return null;
  }
  let children: Array<React.ReactNode> | undefined = undefined;
  if (isHtmlElement(element)) {
    children = element.elements?.map((node, index) => (
      <RichTextElement
        key={index}
        element={node}
        blocks={blocks}
        renderBlock={renderBlock}
        renderNode={renderNode}
        meta={{
          ancestor: element.tag,
          previous: element.elements?.[index - 1]?.tag,
          next: element.elements?.[index + 1]?.tag,
        }}
      />
    ));
    if (children?.length === 0) {
      children = undefined;
    }

    const {
      route,
      style,
      class: className,
      ...attributes
    } = element.attributes;
    const defaultAttributes = htmlAttributes[element.tag];
    if (element.tag === "a" && route?.path) {
      attributes.href = route?.path;
    }

    if (className) {
      if (defaultAttributes?.className) {
        // Merge the default class with the class attribute
        attributes.className = `${defaultAttributes.className} ${className}`;
      } else {
        attributes.className = className;
      }
    }

    if (typeof style === "string") {
      attributes.style = parseStyle(style);
    }

    if (renderNode) {
      const output = renderNode({
        // biome-ignore lint/suspicious/noExplicitAny: Avoid complicated TypeScript logic by using any. The type will be corrected in the implementation.
        tag: element.tag as any,
        attributes: {
          ...defaultAttributes,
          ...attributes,
        } as Record<string, unknown>,
        children,
        route,
        meta: meta || {},
      });

      if (output !== undefined) {
        // If we got a valid output from the renderElement function, we return it
        // `null` we will render nothing, but `undefined` fallback to the default element
        return output;
      }
    }

    return React.createElement(
      element.tag,
      htmlAttributes[element.tag]
        ? { ...defaultAttributes, ...attributes }
        : attributes,
      children,
    );
  }
  return undefined;
}

/**
 * Component for rendering a rich text component
 */
export function UmbracoRichText(props: RichTextProps) {
  const rootElement = props.data;
  if (isRootElement(rootElement)) {
    return (
      <>
        {rootElement.elements?.map((element, index) => (
          <RichTextElement
            key={index}
            element={element}
            blocks={rootElement.blocks}
            renderBlock={props.renderBlock}
            renderNode={props.renderNode}
            htmlAttributes={props.htmlAttributes}
            meta={undefined}
          />
        ))}
      </>
    );
  }

  // If the element is not a root element, we return null
  return null;
}
