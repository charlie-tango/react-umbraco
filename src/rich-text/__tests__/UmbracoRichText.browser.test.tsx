import { render } from "vitest-browser-react";
import type { RichTextElementModel } from "../RichTextTypes";
import { UmbracoRichText } from "../UmbracoRichText";
import fixture from "./__fixtures__/UmbracoRichText.fixture.json";

/**
 * Creates nested RichTextElementModel objects with the given tags.
 * @param {...string} tags - The tags to create nested elements with.
 * @returns {RichTextElementModel[]} - The nested RichTextElementModel object.
 */
function nestedElements(...tags: string[]): RichTextElementModel[] {
  return [
    tags.reduceRight<RichTextElementModel>(
      (acc, tag) => {
        return {
          tag,
          attributes: {},
          elements: [acc],
        } as RichTextElementModel;
      },
      { tag: "#text", attributes: {}, text: "text" } as RichTextElementModel,
    ),
  ];
}

it("should not render without #root tag in the root element ", () => {
  const { container } = render(
    <UmbracoRichText data={{ tag: "div", attributes: {} }} />,
  );
  expect(container.innerHTML).toEqual("");
});

it("should not render if the root element is empty", () => {
  const { container } = render(<UmbracoRichText data={{ tag: "#root" }} />);
  expect(container.innerHTML).toEqual("");
});

it("should render nested elements correctly", () => {
  const { container } = render(
    <UmbracoRichText
      data={{
        tag: "#root",
        elements: nestedElements("table", "thead", "tr", "th"),
      }}
    />,
  );
  const table = container.firstChild as HTMLTableElement;
  expect(table.tagName).toBe("TABLE");
  const thead = table.firstChild as HTMLTableSectionElement;
  expect(thead.tagName).toBe("THEAD");
  const tr = thead.firstChild as HTMLTableRowElement;
  expect(tr.tagName).toBe("TR");
  const th = tr.firstChild as HTMLTableCellElement;
  expect(th.tagName).toBe("TH");
  expect(th.textContent).toBe("text");
});

it("should decode and render text correctly", () => {
  const { container } = render(
    <UmbracoRichText
      data={{
        tag: "#root",
        elements: [
          {
            tag: "#text",
            text: "&lt; &gt; &quot; &apos; &amp; &#169; &#8710;",
          },
        ],
      }}
    />,
  );
  expect(container).toHaveTextContent("< > \" ' & © ∆");
});

it("should throw an error when rendering a block element without renderBlock prop ", () => {
  // don't log errors to the console when testing for thrown errors
  vi.spyOn(console, "error").mockImplementation(() => {});
  expect(() =>
    render(
      <UmbracoRichText
        data={{
          tag: "#root",
          elements: [
            {
              tag: "umb-rte-block",
              attributes: { "content-id": "1" },
              elements: [],
            },
          ],
        }}
      />,
    ),
  ).toThrowError();
});

it("should render a block element using the renderBlock prop", () => {
  const screen = render(
    <UmbracoRichText
      data={{
        tag: "#root",
        blocks: [{ content: { id: "1", properties: {} } }],
        elements: [
          {
            tag: "umb-rte-block",
            attributes: { "content-id": "1" },
            elements: [],
          },
        ],
      }}
      renderBlock={(block) => (
        <div data-testid="umb-rte-block">{block?.content?.id}</div>
      )}
    />,
  );

  expect(
    screen.getByTestId("umb-rte-block").element().outerHTML,
  ).toMatchInlineSnapshot(`"<div data-testid="umb-rte-block">1</div>"`);
});

it("should render an inline block element using the renderBlock prop", () => {
  const screen = render(
    <UmbracoRichText
      data={{
        tag: "#root",
        blocks: [{ content: { id: "1", properties: {} } }],
        elements: [
          {
            tag: "umb-rte-block-inline",
            attributes: { "content-id": "1" },
            elements: [],
          },
        ],
      }}
      renderBlock={(block) => (
        <div data-testid="umb-rte-block-inline">{block?.content?.id}</div>
      )}
    />,
  );

  expect(
    screen.getByTestId("umb-rte-block-inline").element().outerHTML,
  ).toMatchInlineSnapshot(`"<div data-testid="umb-rte-block-inline">1</div>"`);
});

it("should handle href attribute on links", () => {
  const screen = render(
    <UmbracoRichText
      data={{
        tag: "#root",
        elements: [
          {
            tag: "a",
            attributes: {
              href: "/path",
            },
            elements: [{ tag: "#text", text: "text" }],
          },
        ],
      }}
    />,
  );
  const anchor = screen.getByRole("link").element();
  expect(anchor).toHaveAttribute("href", "/path");
  expect(anchor).not.toHaveAttribute("route");
});

it("should handle route-specific path attribute for the anchor href", () => {
  const screen = render(
    <UmbracoRichText
      data={{
        tag: "#root",
        elements: [
          {
            tag: "a",
            attributes: {
              route: { path: "/path", startItem: { id: "1", path: "/path" } },
            },
            elements: [{ tag: "#text", text: "text" }],
          },
        ],
      }}
    />,
  );
  const anchor = screen.getByRole("link").element();
  expect(anchor).toHaveAttribute("href", "/path");
  expect(anchor).not.toHaveAttribute("route");
});

it("should correctly map the class attribute to className", () => {
  const screen = render(
    <UmbracoRichText
      data={{
        tag: "#root",
        elements: [
          {
            tag: "input",
            attributes: { class: "example-class" },
          },
        ],
      }}
    />,
  );
  const input = screen.getByRole("textbox").element();
  expect(input).toHaveClass("example-class");
});

it("should properly convert style string to a CSSProperties object and pass it as the className prop to component", () => {
  const screen = render(
    <UmbracoRichText
      data={{
        tag: "#root",
        elements: [
          {
            tag: "input",
            attributes: { style: "color: red; background-color: blue" },
          },
        ],
      }}
    />,
  );
  const input = screen.getByRole("textbox").element();
  expect(input).toHaveAttribute("style", "color: red; background-color: blue;");
});

it("should override default node rendering with custom renderNode prop", () => {
  const screen = render(
    <UmbracoRichText
      data={{
        tag: "#root",
        elements: [
          {
            tag: "div",
            attributes: {},
          },
        ],
      }}
      renderNode={(node) => <div data-testid="custom-node">{node.tag}</div>}
    />,
  );
  const customNode = screen.getByTestId("custom-node");
  expect(customNode.element().tagName).toBe("DIV");
});

it("should not render node if renderNode prop returns null", () => {
  const screen = render(
    <UmbracoRichText
      data={{
        tag: "#root",
        elements: [
          {
            tag: "div",
            attributes: {
              "data-testid": "default-rendering",
            },
          },
        ],
      }}
      renderNode={() => null}
    />,
  );
  expect(() => screen.getByTestId("default-rendering").element()).to.throw();
});

it("should render default node if renderNode prop returns undefined", () => {
  const screen = render(
    <UmbracoRichText
      data={{
        tag: "#root",
        elements: [
          {
            tag: "div",
            attributes: {
              "data-testid": "default-rendering",
            },
          },
        ],
      }}
      renderNode={() => undefined}
    />,
  );
  expect(screen.getByTestId("default-rendering")).not.toBeNull();
});

it("should render fixture content correctly", () => {
  const screen = render(
    <UmbracoRichText
      // biome-ignore lint/suspicious/noExplicitAny:
      data={fixture as any}
      renderBlock={() => <div data-testid="block" />}
    />,
  );
  expect(screen.container.innerHTML).toMatchSnapshot();
});

it("should handle default attributes for elements", () => {
  const screen = render(
    <UmbracoRichText
      data={{
        tag: "#root",
        elements: [
          {
            tag: "h1",
            attributes: {},
            elements: [
              {
                text: "Welcome",
                tag: "#text",
              },
            ],
          },
          {
            tag: "p",
            attributes: {},
            elements: [
              {
                text: "What follows from here is just a bunch of absolute nonsense I've written to dogfood the plugin itself.",
                tag: "#text",
              },
            ],
          },
          {
            tag: "h2",
            attributes: { class: "pre-styled" },
            elements: [
              {
                text: "h2 is already styled",
                tag: "#text",
              },
            ],
          },
        ],
      }}
      htmlAttributes={{
        p: { className: "mb-4" },
        h1: { className: "text-2xl", style: { color: "red" } },
        h2: { className: "text-1xl" },
      }}
    />,
  );

  const [h1, h2] = screen
    .getByRole("heading")
    .all()
    .map((heading) => heading.element());
  expect(h1).toHaveClass("text-2xl");
  expect(h1).toHaveAttribute("style", "color: red;");
  expect(h2).toHaveClass("text-1xl"); //
  expect(h2).toHaveClass("pre-styled"); // `h2` has a class by default. It should be preserved

  const paragraph = screen.getByRole("paragraph");
  expect(paragraph.element()).toHaveClass("mb-4");
});

it("should handle default attributes for with renderNode", () => {
  const screen = render(
    <UmbracoRichText
      data={{
        tag: "#root",
        elements: [
          {
            tag: "p",
            attributes: {},
            elements: [
              {
                text: "What follows from here is just a bunch of absolute nonsense I've written to dogfood the plugin itself.",
                tag: "#text",
              },
            ],
          },
        ],
      }}
      htmlAttributes={{
        p: { className: "mb-4" },
        h1: { className: "text-2xl", style: { color: "red" } },
        h2: { className: "text-1xl" },
      }}
      renderNode={(node) => {
        if (node.tag === "p") {
          return (
            <p
              {...node.attributes}
              className={[node.attributes.className, "font-medium"]
                .filter(Boolean)
                .join(" ")}
            >
              {node.children}
            </p>
          );
        }
      }}
    />,
  );

  const paragraph = screen.getByRole("paragraph").element();
  expect(paragraph).toHaveClass("mb-4");
  expect(paragraph).toHaveClass("mb-4 font-medium");
});
