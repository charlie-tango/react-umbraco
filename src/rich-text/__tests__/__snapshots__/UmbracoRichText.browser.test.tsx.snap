// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render fixture content correctly 1`] = `
"<h2>What to expect from here on out</h2><div data-testid="block"></div><p><div data-testid="block"></div></p><p>What follows from here is just a bunch of absolute nonsense I've written to dogfood the plugin itself. It includes every sensible typographic element I could think of, like <strong>bold text</strong>, unordered lists, ordered lists, code blocks, block quotes, <em>and even italics</em>.</p><p>It's important to cover all of these use cases for a few reasons:</p><ol><li>We want everything to look good out of the box.</li><li>Really just the first reason, that's the whole point of the plugin.</li><li>Here's a third pretend reason though a list with three items looks more realistic than a list with two items.</li></ol><p>Now we're going to try out another header style.</p><h3>Typography should be easy</h3><p>So that's a header for you — with any luck if we've done our job correctly that will look pretty reasonable.</p><p>Something a wise person once told me about typography is:</p><blockquote><p>Typography is pretty important if you don't want your stuff to look like trash. Make it good then it won't be bad.</p></blockquote><p>It's probably important that images look okay here by default as well:</p><figure><img alt="" src="https://images.unsplash.com/photo-1556740758-90de374c12ad?ixlib=rb-1.2.1&amp;amp;ixid=eyJhcHBfaWQiOjEyMDd9&amp;amp;auto=format&amp;amp;fit=crop&amp;amp;w=1000&amp;amp;q=80"><figcaption>Contrary to popular belief, Lorem Ipsum is not simply random text. It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years old.</figcaption></figure><p>Now I'm going to show you an example of an unordered list to make sure that looks good, too:</p><ul><li>So here is the first item in this list.</li><li>In this example we're keeping the items short.</li><li>Later, we'll use longer, more complex list items.</li></ul><p>And that's the end of this section.</p><h2>What if we stack headings?</h2><h3>We should make sure that looks good, too.</h3><p>Sometimes you have headings directly underneath each other. In those cases you often have to undo the top margin on the second heading because it usually looks better for the headings to be closer together than a paragraph followed by a heading should be.</p><h3>When a heading comes after a paragraph …</h3><p>When a heading comes after a paragraph, we need a bit more space, like I already mentioned above. Now let's see what a more complex list would look like.</p><ul><li><p><strong>I often do this thing where list items have headings.</strong></p><p>For some reason I think this looks cool which is unfortunate because it's pretty annoying to get the styles right.</p><p>I often have two or three paragraphs in these list items, too, so the hard part is getting the spacing between the paragraphs, list item heading, and separate list items to all make sense. Pretty tough honestly, you could make a strong argument that you just shouldn't write this way.</p></li><li><p><strong>Since this is a list, I need at least two items.</strong></p><p>I explained what I'm doing already in the previous list item, but a list wouldn't be a list if it only had one item, and we really want this to look realistic. That's why I've added this second list item so I actually have something to look at when writing the styles.</p></li><li><p><strong>It's not a bad idea to add a third item either.</strong></p><p>I think it probably would've been fine to just use two items but three is definitely not worse, and since I seem to be having no trouble making up arbitrary things to type, I might as well include it. I'm going to press Enter now.</p></li></ul><p>After this sort of list I usually have a closing statement or paragraph, because it kinda looks weird jumping right to a heading.</p><h2>Code should look okay by default.</h2><p>I think most people are going to use <a href="https://highlightjs.org/">highlight.js</a> or <a href="https://prismjs.com/">Prism</a> or something if they want to style their code blocks but it wouldn't hurt to make them look <em>okay</em> out of the box, even with no syntax highlighting.</p><p>Here's what a default tailwind.config.js file looks like at the time of writing:</p><pre>module.exports = {
  purge: [],
  theme: {
    extend: {},
  },
  variants: {},
  plugins: [],
}
</pre><p>Hopefully that looks good enough to you.</p><h3>What about nested lists?</h3><p>Nested lists basically always look bad which is why editors like Medium don't even let you do it, but I guess since some of you goofballs are going to do it we have to carry the burden of at least making it work.</p><ol><li><strong>Nested lists are rarely a good idea.</strong><ul><li>You might feel like you are being really "organized" or something but you are just creating a gross shape on the screen that is hard to read.</li><li>Nested navigation in UIs is a bad idea too, keep things as flat as possible.</li><li>Nesting tons of folders in your source code is also not helpful.</li></ul></li><li><strong>Since we need to have more items, here's another one.</strong><ul><li>I'm not sure if we'll bother styling more than two levels deep.</li><li>Two is already too much, three is guaranteed to be a bad idea.</li><li>If you nest four levels deep you belong in prison.</li></ul></li><li><strong>Two items isn't really a list, three is good though.</strong><ul><li>Again please don't nest lists if you want people to actually read your content.</li><li>Nobody wants to look at this.</li><li>I'm upset that we even have to bother styling this.</li></ul></li></ol><p>The most annoying thing about lists in Markdown is that &lt;li&gt; elements aren't given a child &lt;p&gt; tag unless there are multiple paragraphs in the list item. That means I have to worry about styling that annoying situation too.</p><ul><li><p><strong>For example, here's another nested list.</strong></p><p>But this time with a second paragraph.</p><ul><li>These list items won't have &lt;p&gt; tags</li><li>Because they are only one line each</li></ul></li><li><p><strong>But in this second top-level list item, they will.</strong></p><p>This is especially annoying because of the spacing on this paragraph.</p><ul><li><p>As you can see here, because I've added a second line, this list item now has a &lt;p&gt; tag.</p><p>This is the second line I'm talking about by the way.</p></li><li><p>Finally here's another list item so it's more like a list.</p></li></ul></li><li><p>A closing list item, but with no nested list, because why not?</p></li></ul><p>And finally a sentence to close off this section.</p><h2>We didn't forget about description lists</h2><p>Well, that's not exactly true, we first released this plugin back in 2020 and it took three years before we added description lists. But they're here now, so let's just be happy about that…okay? They can be great for things like FAQs.</p><dl><dt>Why do you never see elephants hiding in trees?</dt><dd>Because they're so good at it. Lorem ipsum dolor sit amet consectetur adipisicing elit. Quas cupiditate laboriosam fugiat.</dd><dt>What do you call someone with no body and no nose?</dt><dd>Nobody knows. Lorem ipsum dolor sit amet consectetur adipisicing elit. Culpa, voluptas ipsa quia excepturi, quibusdam natus exercitationem sapiente tempore labore voluptatem.</dd><dt>Why can't you hear a pterodactyl go to the bathroom?</dt><dd>Because the pee is silent. Lorem ipsum dolor sit amet, consectetur adipisicing elit. Ipsam, quas voluptatibus ex culpa ipsum, aspernatur blanditiis fugiat ullam magnam suscipit deserunt illum natus facilis atque vero consequatur! Quisquam, debitis error.</dd></dl><h2>There are other elements we need to style</h2><p>I almost forgot to mention links, like <a href="https://tailwindcss.com">this link to the Tailwind CSS website</a>. We almost made them blue but that's so yesterday, so we went with dark gray, feels edgier.</p><p>We even included table styles, check it out:</p><table border="0"><thead><tr><th>Wrestler</th><th>Origin</th><th>Finisher</th></tr></thead><tbody><tr><td>Bret "The Hitman" Hart</td><td>Calgary, AB</td><td>Sharpshooter</td></tr><tr><td>Stone Cold Steve Austin</td><td>Austin, TX</td><td>Stone Cold Stunner</td></tr><tr><td>Randy Savage</td><td>Sarasota, FL</td><td>Elbow Drop</td></tr><tr><td>Vader</td><td>Boulder, CO</td><td>Vader Bomb</td></tr><tr><td>Razor Ramon</td><td>Chuluota, FL</td><td>Razor's Edge</td></tr></tbody></table><p>We also need to make sure inline code looks good, like if I wanted to talk about &lt;span&gt; elements or tell you the good news about @tailwindcss/typography.</p><h3>Sometimes I even use code in headings</h3><p>Even though it's probably a bad idea, and historically I've had a hard time making it look good. This <em>"wrap the code blocks in backticks"</em> trick works pretty well though really.</p><p>Another thing I've done in the past is put a code tag inside of a link, like if I wanted to tell you about the <a href="https://github.com/tailwindcss/docs">tailwindcss/docs</a> repository. I don't love that there is an underline below the backticks but it is absolutely not worth the madness it would require to avoid it.</p><h4>We haven't used an h4 yet</h4><p>But now we have. Please don't use h5 or h6 in your content, Medium only supports two heading levels for a reason, you animals. I honestly considered using a before pseudo-element to scream at you if you use an h5 or h6.</p><p>We don't style them at all out of the box because h4 elements are already so small that they are the same size as the body copy. What are we supposed to do with an h5, make it <em>smaller</em> than the body copy? No thanks.</p><h3>We still need to think about stacked headings though.</h3><h4>Let's make sure we don't screw that up with h4 elements, either.</h4><p>Phew, with any luck we have styled the headings above this text and they look pretty good.</p><p>Let's add a closing paragraph here so things end with a decently sized block of text. I can't explain why I want things to end that way but I have to assume it's because I think things will look weird or unbalanced if there is a heading too close to the end of the document.</p><p>What I've written here is probably long enough, but adding this final sentence can't hurt.</p><h3>Umbraco&nbsp;</h3><p>We shouldn't forget about&nbsp;<a title="Contentpage" href="/contentpage/">internal links</a> to Umbraco content.</p><p>It's also interesting to see how Umbraco outputs images.</p><p>&nbsp;</p><figure><img alt="The harbour at night" width="604" height="353" caption="The harbour at night" src="https://diversa-dk.euwest01.umbraco.io/media/mhqclvbv/hero-image.png"><figcaption>The harbour at night</figcaption></figure><p>&nbsp;</p>"
`;
